### PR TITLE
feat: improve workload reconcile with per-resource ProgressGroup tracking

### PR DESCRIPTION
**Pull request by devantler**
## Summary

Rewrites the `ksail workload reconcile` command output to provide per-resource progress tracking using the existing `ProgressGroup` pattern, topological dependency ordering for Flux kustomizations, and actionable error messages on failure.

**Before** — sequential activity notifications with no per-resource visibility:
```
🔄 Trigger Reconciliation...
► triggering flux oci repository reconciliation
► waiting for flux oci repository to be ready
► triggering flux kustomization reconciliation
► waiting for flux kustomization to reconcile
► waiting for all flux kustomizations to reconcile
✔ reconciliation completed
```

**After** — per-resource ProgressGroup tracking in topological dependency order:
```
🔄 Trigger Reconciliation...
► reconciling oci source...
✔ flux-system reconciled [1.2s]
► reconciling kustomizations...
✔ flux-system reconciled [3.1s]
✔ infra reconciled [5.2s]
✔ apps reconciled [8.4s]
✔ reconciliation completed [12.3s]
```

### Core
- **`pkg/notify/progress.go`**: Add `ReconcilingLabels()` for reconciliation progress labels; skip title line when title is empty (allows sub-phase ProgressGroups)
- **`pkg/client/flux/reconciler.go`**: Extend `KustomizationInfo` with `DependsOn` field; add `ListKustomizations()`, `CheckNamedKustomizationReady()`, `parseDependsOn()`
- **`pkg/client/argocd/reconciler.go`**: Add `ApplicationInfo` struct, `ListApplications()`, `CheckNamedApplicationReady()`
- **`pkg/cli/cmd/workload/workload.go`**: Rewrite `reconcileFlux()` and `reconcileArgoCD()` with ProgressGroup per-resource tracking; add `topologicalSortKustomizations()` (Kahn's algorithm); add actionable timeout/failure error messages

### Tests
- **`pkg/notify/progress_test.go`**: Tests for `ReconcilingLabels()` and empty-title ProgressGroup
- **`pkg/client/flux/reconciler_test.go`**: Tests for `ListKustomizations()` (deps parsing, edge cases), `CheckNamedKustomizationReady()` (ready, transient, permanent failure, stalled)
- **`pkg/client/argocd/reconciler_test.go`**: Tests for `ListApplications()`, `CheckNamedApplicationReady()` (synced+healthy, degraded, operation failures, source errors)
- **`pkg/cli/cmd/workload/workload_test.go`**: Tests for `topologicalSortKustomizations()` (empty, single, linear chain, diamond, cycle, external deps)
- **`pkg/cli/cmd/workload/export_test.go`**: Export seam for topological sort
- **`pkg/cli/cmd/workload/__snapshots__/workload_test.snap`**: Updated help snapshot

## Design decisions

- **Display-only topological sort**: Flux's controller handles actual dependency-driven reconciliation. We trigger the root kustomization and monitor all kustomizations — the sort is purely for display ordering
- **Graceful cycle handling**: If cycles exist in kustomization dependencies, remaining items are appended in original order
- **Dead code removal**: Previously exported wait-style APIs (`WaitForKustomizationReady`, `WaitForAllKustomizationsReady`, `WaitForApplicationReady`) had zero callers and were removed to satisfy the CI deadcode check. No external consumers of these internal modules exist
- **Shared deadline preserved**: Single timeout across all reconciliation phases

### CODE REVIEW

File: pkg/notify/progress_test.go [hunk 1]
```diff
@@ -652,6 +652,65 @@ func TestProgressGroup_InstallingLabels(t *testing.T) {
...
+		t.Errorf("expected no error, got: %v", err)
+	}
+
+	output := buf.String()
+
+	// The title line format is "<emoji> <title>...\n". When both are empty,
+	// this line must be absent.
+	if strings.Contains(output, " ...\n") {
+		t.Errorf("expected no title line when title is empty, got: %q", output)
+	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/notify/progress_test.go:720]: Good  the original assertion was weak. Fixed: now splits output into lines and asserts the first line is not a title-format line.

File: pkg/cli/cmd/workload/workload.go [hunk 2]
```diff
@@ -2099,137 +2102,359 @@ func executeReconciliation(
...
+		writer,
+		notify.WithLabels(notify.ReconcilingLabels()),
+		notify.WithTimer(outputTimer),
 	)
 
-	err = reconciler.WaitForOCIRepositoryReady(cmd.Context(), timeout)
+	err = ociGroup.Run(cmd.Context(), notify.ProgressTask{
+		Name: "flux-system",
+		Fn: func(ctx context.Context) error {
+			return fluxReconciler.WaitForOCIRepositoryReady(ctx, timeout)
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: now passes `time.Until(deadline)` to `WaitForOCIRepositoryReady` so both phases share the deadline correctly.

File: pkg/cli/cmd/workload/workload.go [hunk 3]
```diff
@@ -2099,137 +2102,359 @@ func executeReconciliation(
...
+		name := queue[0]
+		queue = queue[1:]
+		sorted = append(sorted, byName[name])
+
+		for _, ks := range kustomizations {
+			if !slices.Contains(ks.DependsOn, name) {
+				continue
+			}
+
+			inDegree[ks.Name]--
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: Fixed by de-duplicating dependencies per kustomization during in-degree counting. Added a test case for duplicate deps.

File: pkg/cli/cmd/workload/workload.go [hunk 4]
```diff
@@ -2099,137 +2102,359 @@ func executeReconciliation(
...
+	defer ticker.Stop()
+
+	var lastStatus string
+
+	for {
+		ready, status, err := fluxReconciler.CheckNamedKustomizationReady(timeoutCtx, name)
+		if err != nil {
+			if reconciler.IsContextError(err) {
+				return kustomizationReadinessTimeoutError(name, lastStatus)
+			}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: removed the name prefix from poll error returns since ProgressGroup already wraps errors as `<task name>: <err>`. Timeout errors now omit the resource name in the message prefix and instead include it in the actionable CLI command suggestion.

File: pkg/cli/cmd/workload/workload.go [hunk 5]
```diff
@@ -2099,137 +2102,359 @@ func executeReconciliation(
...
+	sorted := topologicalSortKustomizations(kustomizations)
+
+	tasks := make([]notify.ProgressTask, 0, len(sorted))
+	for _, ks := range sorted {
+		name := ks.Name
+		tasks = append(tasks, notify.ProgressTask{
+			Name: name,
+			Fn: func(ctx context.Context) error {
+				return pollUntilKustomizationReady(ctx, fluxReconciler, name, timeout)
+			},
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go:2196]: Added `notify.WithConcurrency(reconcileConcurrency)` (cap of 5) to both the Flux kustomization and ArgoCD application ProgressGroups, matching the existing pattern used by validation.

File: pkg/cli/cmd/workload/workload.go [hunk 6]
```diff
@@ -2099,137 +2102,359 @@ func executeReconciliation(
...
+	}
+
+	tasks := make([]notify.ProgressTask, 0, len(apps))
+	for _, app := range apps {
+		name := app.Name
+		tasks = append(tasks, notify.ProgressTask{
+			Name: name,
+			Fn: func(ctx context.Context) error {
+				return pollUntilApplicationReady(ctx, argoReconciler, name, timeout)
+			},
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go:2397]: Addressed together with the Flux concurrency  both ProgressGroups now use `WithConcurrency(5)`.

File: pkg/cli/cmd/workload/workload.go [hunk 7]
```diff
@@ -2099,137 +2103,364 @@ func executeReconciliation(
...
+
+	tasks := make([]notify.ProgressTask, 0, len(sorted))
+	for _, ks := range sorted {
+		name := ks.Name
+		tasks = append(tasks, notify.ProgressTask{
+			Name: name,
+			Fn: func(ctx context.Context) error {
+				return pollUntilKustomizationReady(ctx, fluxReconciler, name, timeout)
+			},
+		})
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: Fixed: replaced per-task `context.WithTimeout(ctx, timeout)` with a single shared deadline context (`ksCtx`) created in `reconcileFlux()` and passed to `ProgressGroup.Run()`. Each task now respects the shared deadline via `ctx.Done()`, so total reconcile time is bounded regardless of concurrency.

File: pkg/cli/cmd/workload/workload.go [hunk 8]
```diff
@@ -2099,137 +2103,364 @@ func executeReconciliation(
...
+		notify.WithCountLabel("applications"),
+		notify.WithConcurrency(reconcileConcurrency),
 	)
 
-	err = reconciler.WaitForApplicationReady(cmd.Context(), timeout)
+	err = appGroup.Run(cmd.Context(), tasks...)
 	if err != nil {
-		return fmt.Errorf("wait for argocd application ready: %w", err)
+		return fmt.Errorf("reconcile argocd applications: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: Fixed: same approach as Flux. Created a single `deadlineCtx` with `context.WithTimeout(cmd.Context(), timeout)` in `reconcileArgoCD()` and pass it to `appGroup.Run()`. Each `pollUntilApplicationReady` task uses `ctx.Done()` from that shared context.

File: pkg/notify/progress_test.go [hunk 9]
```diff
@@ -652,6 +652,70 @@ func TestProgressGroup_InstallingLabels(t *testing.T) {
...
+	lines := strings.Split(strings.TrimSuffix(output, "\n"), "\n")
+	if len(lines) == 0 || lines[0] == "" {
+		t.Fatalf("expected task output, got: %q", output)
+	}
+
+	// With an empty title, the first rendered line should be the task output,
+	// not a title line such as " ...".
+	if strings.HasPrefix(lines[0], " ...") || strings.TrimSpace(lines[0]) == "..." {
+		t.Errorf("expected no title line when title is empty, got first line %q in output %q", lines[0], output)
+	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/notify/progress_test.go]: Fixed: added assertions for the `"\u25ba ..."` prefix pattern and a positive check that the first line contains the task name `"my-task"`.

File: pkg/client/flux/reconciler.go [hunk 10]
```diff
@@ -185,122 +186,62 @@ func (r *Reconciler) ListKustomizationPaths(
...
+	kustomization, err := client.Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return false, "", fmt.Errorf("get flux kustomization %q: %w", name, err)
 	}
 
-	if len(kustomizations.Items) == 0 {
-		return true, "", nil
-	}
+	return checkKustomizationStatus(kustomization)
+}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/flux/reconciler.go:220]: These methods have zero callers anywhere in the codebase (confirmed by `deadcode -test ./...`). Re-adding them as wrappers would immediately fail CI because the deadcode analysis is a required check. Since `pkg/client/flux` is internal to this application (no external consumers of the module), this is not a breaking change in practice. Updated the PR description to remove the backward-compatibility claim.

File: pkg/client/argocd/reconciler.go [hunk 11]
```diff
@@ -102,95 +101,69 @@ func (r *Reconciler) TriggerRefresh(ctx context.Context, hardRefresh bool) error
...
 
-	// Check for error conditions
 	err = r.checkConditions(app)
 	if err != nil {
 		return false, err
 	}
 
-	// Check if synced and healthy
 	return isApplicationSynced(app), nil
 }
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/argocd/reconciler.go:155]: Same reasoning as the Flux thread: `WaitForApplicationReady` had zero callers (verified by `deadcode -test ./...`). Re-adding it would fail the required deadcode CI check. No external consumers of this module exist.

File: pkg/client/flux/reconciler.go [hunk 12]
```diff
@@ -185,122 +186,62 @@ func (r *Reconciler) ListKustomizationPaths(
...
 
-	if len(notReady) == 0 {
-		return true, "", nil
-	}
-
-	sort.Strings(notReady)
-
-	return false, "not ready: " + strings.Join(notReady, ", "), nil
+	return names
 }
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/flux/reconciler.go]: Fixed: `parseDependsOn` now checks the `namespace` field and skips entries where `namespace` is non-empty and differs from `DefaultNamespace` ("flux-system"). Cross-namespace deps are treated as external and excluded from topological ordering.

File: pkg/cli/cmd/workload/workload.go [hunk 13]
```diff
@@ -2099,137 +2103,366 @@ func executeReconciliation(
...
+		return fmt.Errorf("list argocd applications: %w", err)
+	}
+
+	if len(apps) == 0 {
+		return nil
+	}
+
+	deadlineCtx, deadlineCancel := context.WithTimeout(cmd.Context(), timeout)
+	defer deadlineCancel()
+
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go]: Fixed: moved `deadlineCtx` creation before both `TriggerRefresh` and `ListApplications` so the entire ArgoCD reconcile phase (including the list call and refresh trigger) respects the configured timeout.

File: pkg/cli/cmd/workload/workload.go [hunk 14]
```diff
@@ -2099,137 +2103,366 @@ func executeReconciliation(
...
-		cmd.OutOrStdout(),
-	)
+	// Sub-phase 2: Kustomization reconciliation with per-resource tracking
+	writeActivityNotification("reconciling kustomizations...", outputTimer, writer)
 
-	err = reconciler.TriggerKustomizationReconciliation(cmd.Context())
+	err = fluxReconciler.TriggerKustomizationReconciliation(cmd.Context())
 	if err != nil {
 		return fmt.Errorf("trigger kustomization reconciliation: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go:2158]: Fixed: `reconcileFlux` now creates a single `deadlineCtx` upfront and uses it for both trigger calls (`TriggerOCIRepositoryReconciliation`, `TriggerKustomizationReconciliation`), the OCI wait, and the kustomization ProgressGroup phase. The entire Flux reconcile is now bounded by the configured timeout.

File: pkg/cli/cmd/workload/workload.go [hunk 15]
```diff
@@ -2099,137 +2103,351 @@ func executeReconciliation(
...
+		notify.WithTimer(outputTimer),
 	)
 
-	err = reconciler.WaitForOCIRepositoryReady(cmd.Context(), timeout)
+	err = ociGroup.Run(deadlineCtx, notify.ProgressTask{
+		Name: "flux-system",
+		Fn: func(ctx context.Context) error {
+			return fluxReconciler.WaitForOCIRepositoryReady(ctx, time.Until(deadline))
+		},
+	})
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go:2147]: Fixed: added `WithAppendOnly()` to the OCI progress group for consistent output with the kustomization and application groups.

File: pkg/cli/cmd/workload/workload.go [hunk 16]
```diff
@@ -2099,137 +2103,350 @@ func executeReconciliation(
...
+		if ready {
+			return nil
+		}
+
+		lastStatus = status
+
+		select {
+		case <-ctx.Done():
+			return kustomizationReadinessTimeoutError(name, lastStatus)
+		case <-ticker.C:
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/workload/workload.go:2262]: Fixed: both poll functions now check `ctx.Err() == context.Canceled` and propagate the cancellation directly. The actionable timeout message with inspect guidance is only shown on `context.DeadlineExceeded`.
